Dev/production doc (#824)
AMD-AGI/Primus: production_docs/02-user-guide/pretraining.md

> # Pretraining Workflows

Reviewer: We say Post-training. Let's be consistent with whether there's a - before training.

> Primus is a YAML-driven training stack for AMD GPUs. You select a **backend** (Megatron-LM, TorchTitan, JAX MaxText, Megatron Bridge), point `train pretrain` at an **experiment YAML**, and launch with the unified CLI (`runner/primus-cli`) in **direct**, **container**, or **Slurm** mode. See [CLI Reference](cli-reference.md) and [Configuration System](configuration-system.md).

Reviewer: ```suggestion
Primus is a YAML-driven training stack for AMD GPUs. You select a **backend** (Megatron-LM, TorchTitan, JAX MaxText, Megatron Bridge), point `train pretrain` at an **experiment YAML**, and launch Primus with the unified CLI (`runner/primus-cli`) in **direct**, **container**, or **Slurm** mode. See [CLI Reference](cli-reference.md) and [Configuration System](configuration-system.md).
```

Reviewer: Not sure why we call it "experiment YAML" -- it can be a real training, right? How about point .... at a configuration YAML?

Reviewer: But if there's a change for this, make sure it is applied to ALL files ALL instances.

> **Pretraining has two docs — pick the one that matches your question:**
> - **Use this doc** to understand the *why*: how backends work, YAML structure and inheritance, parallelism vocabulary, and the full per-backend config inventory.
> - **Use [Backend Training Recipes](training-recipes.md)** when you already know *what* you want to run and just need the exact, GPU-arch-specific command.

> 

Reviewer: This is a bit confusing. Change the above to a simple paragraph like the one below (can keep the indent): 

This sections help you to understand concepts related to Primus workflow (how backends work, YAML structure and inheritance, parallelism vocabulary, the full per-backend configuration inventory, etc.). If you already have a good understanding on the concepts just need the specific commands to run your training with Primus, see [Backend Training Recipes](training-recipes.md)**.

> **Related documentation**

Reviewer: Move this to the end of the page, and use the same format that is used on other pages (no tables, just bullets).

| Topic | Location |
| --- | --- |
| Launcher usage | [CLI Reference](cli-reference.md) |
| YAML merge rules | [Configuration System](configuration-system.md) |
| Backend knobs | [Megatron parameters](../03-configuration-reference/megatron-parameters.md), [TorchTitan parameters](../03-configuration-reference/torchtitan-parameters.md), [MaxText parameters](../03-configuration-reference/maxtext-parameters.md) |

---

## Overview

| Backend | Framework | Typical use |
| --- | --- | --- |
| Megatron-LM | `framework: megatron` | Large-scale transformer pretraining with Megatron-style parallelism (TP/PP/EP). |
| TorchTitan | `framework: torchtitan` | PyTorch-native scaled training (FSDP / tensor / pipeline / expert parallelism per config). |
| MaxText (JAX) | `framework: maxtext` | JAX/MaxText single- and multi-node runs; parallelism via MaxText `ici_*` / `dcn_*` settings. |
| Megatron Bridge | `framework: megatron_bridge` | Bridge-oriented workflows (configure like other backends; see parameter reference). |

---

## Megatron-LM pretraining

### Quick start (container mode)

> From the repository root, with Docker or Podman available:

Reviewer: ```suggestion
From the root of the clone of the [Primus repository](https://github.com/AMD-AGI/Primus), with Docker or Podman available, the following command starts the training in container mode:
```

```bash
./runner/primus-cli container -- train pretrain \
  --config examples/megatron/configs/MI300X/llama2_7B-BF16-pretrain.yaml
```

This uses the default image from `runner/.primus.yaml` (`rocm/primus:v26.2` unless overridden). The project tree is mounted into the container automatically by `runner/primus-cli-container.sh`.

> ### Example walkthrough: `llama2_7B-BF16-pretrain.yaml`

Reviewer: Move this down to after Example configs -- that makes the flow better and things more understandable.

Reviewer: Also, as this is not a walkthrough of steps, change title to something like : 

### Sample  YAML file (`llama2_7B-BF16-pretrain.yaml`) Explained

Path: `examples/megatron/configs/MI300X/llama2_7B-BF16-pretrain.yaml`

| Section | Role |
| --- | --- |
| `work_group`, `user_name`, `exp_name`, `workspace` | Run identity and output root (supports `${VAR:default}` substitution). |
| `modules.pre_trainer.framework` | `megatron` selects Megatron-LM integration. |
| `config: pre_trainer.yaml` | Module preset under `primus/configs/modules/megatron/`. |
| `model: llama2_7B.yaml` | Model preset under `primus/configs/models/megatron/` (extends `llama2_base.yaml` → …). |
| `overrides` | Experiment-specific training knobs: iterations, batching, LR, **parallelism** (`tensor_model_parallel_size`, `pipeline_model_parallel_size`, `expert_model_parallel_size`), data paths, checkpoints, Primus Turbo flags, etc. |

The sample sets `mock_data: true` and `train_data_path: null` so you can validate the stack without real corpora.

> ### Example configs under `examples/megatron/configs/MI300X/`

Reviewer: ```suggestion
### Example configurations under `examples/megatron/configs/MI300X/`
```

| `qwen2.5_7B-FP8-pretrain.yaml` | `1` | `1` | `1` |
| `qwen3_14B-BF16-pretrain.yaml` | `1` | `1` | `1` |
| `qwen3_14B-FP8-pretrain.yaml` | `1` | `1` | `1` |
| `qwen3_235B_A22B-BF16-pretrain.yaml` | `1` | `1` | `8` |
| `qwen3_235B_A22B-FP8-pretrain.yaml` | `1` | `1` | `8` |
| `qwen3_30B_A3B-BF16-pretrain.yaml` | `${PRIMUS_TP:1}` | `${PRIMUS_PP:1}` | `${PRIMUS_EP:8}` |
| `qwen3_30B_A3B-FP8-pretrain.yaml` | `${PRIMUS_TP:1}` | `${PRIMUS_PP:1}` | `${PRIMUS_EP:8}` |
| `qwen3_32B-BF16-pretrain.yaml` | `1` | `1` | `1` |
| `qwen3_32B-FP8-pretrain.yaml` | `1` | `1` | `1` |
| `qwen3_4B-BF16-pretrain.yaml` | `1` | `1` | `1` |
| `qwen3_4B-FP8-pretrain.yaml` | `1` | `1` | `1` |
| `qwen3_5_35B_A3B-BF16-pretrain.yaml` | `${PRIMUS_TP:1}` | `${PRIMUS_PP:1}` | `${PRIMUS_EP:8}` |
| `qwen3_5_35B_A3B-FP8-pretrain.yaml` | `${PRIMUS_TP:1}` | `${PRIMUS_PP:1}` | `${PRIMUS_EP:8}` |
| `qwen3_8B-BF16-pretrain.yaml` | `1` | `1` | `1` |
| `qwen3_8B-FP8-pretrain.yaml` | `1` | `1` | `1` |
| `zebra_llama_1B-pretrain.yaml` | `1` | `1` | `1` |
| `zebra_llama_3B-pretrain.yaml` | `1` | `1` | `1` |
| `zebra_llama_8B-pretrain.yaml` | `1` | `1` | `1` |

### Mock data versus real data

> - **Mock data:** set `mock_data: true` and leave `train_data_path` / `valid_data_path` empty (as in `llama2_7B-BF16-pretrain.yaml`).  

Reviewer: ```suggestion
- **Mock data:** Set `mock_data: true` and leave `train_data_path` / `valid_data_path` empty (as in `llama2_7B-BF16-pretrain.yaml`).  
```

> - **Real data:** set `mock_data: false` and populate Megatron-compatible data paths (and tokenizer assets) in `overrides`. Use paths visible inside your container mounts.

Reviewer: ```suggestion
- **Real data:** Set `mock_data: false` and populate Megatron-compatible data paths (and tokenizer assets) in `overrides`. Use paths visible inside your container mounts.
```

### Multi-node training with Slurm

```bash
./runner/primus-cli slurm srun -N 4 -p <partition> -- train pretrain \
  --config examples/megatron/configs/MI300X/llama2_7B-BF16-pretrain.yaml
```

`runner/primus-cli-slurm-entry.sh` derives `MASTER_ADDR`, `NNODES`, and `NODE_RANK` from Slurm and forwards them into the container. Align `tensor_model_parallel_size`, `pipeline_model_parallel_size`, and `expert_model_parallel_size` with your cluster width and job size.

---

## TorchTitan pretraining

### Quick start

```bash
./runner/primus-cli container -- train pretrain \
  --config examples/torchtitan/configs/MI300X/llama3.1_8B-BF16-pretrain.yaml
```

### Example walkthrough

Path: `examples/torchtitan/configs/MI300X/llama3.1_8B-BF16-pretrain.yaml`

| Section | Role |
| --- | --- |
| `framework: torchtitan` | Selects the TorchTitan integration. |
| `config: pre_trainer.yaml` | Module preset under `primus/configs/modules/torchtitan/`. |
| `model: llama3.1_8B.yaml` | Model preset under `primus/configs/models/torchtitan/`. |
| `overrides.training`, `lr_scheduler`, `activation_checkpoint`, `primus_turbo` | Run-specific batching, steps, checkpointing, and Turbo options. |

> Some experiments omit an explicit `parallelism:` block; in that case defaults come from the **module and model presets** (`primus/configs/modules/torchtitan/pre_trainer.yaml` and the chosen model YAML). Other examples (for example DeepSeek and Qwen) set `parallelism:` inline with `tensor_parallel_degree`, `pipeline_parallel_degree`, `expert_parallel_degree`, etc.

Reviewer: Again, do we want to use "experiments" or just "configurations"?

> Some experiments omit an explicit `parallelism:` block; in that case defaults come from the **module and model presets** (`primus/configs/modules/torchtitan/pre_trainer.yaml` and the chosen model YAML). Other examples (for example DeepSeek and Qwen) set `parallelism:` inline with `tensor_parallel_degree`, `pipeline_parallel_degree`, `expert_parallel_degree`, etc.

Reviewer: ```suggestion
Some experiments omit an explicit `parallelism:` block; in that case the default values come from the **module and model presets** (`primus/configs/modules/torchtitan/pre_trainer.yaml` and the chosen model YAML). Other examples (for example DeepSeek and Qwen) set `parallelism:` inline with `tensor_parallel_degree`, `pipeline_parallel_degree`, `expert_parallel_degree`, etc.
```

| `llama3.3_70B-pretrain.yaml` | `ici_fsdp_parallelism: 8`, `ici_data_parallelism: 1`, `dcn_fsdp_parallelism: 1`, `dcn_data_parallelism: -1` |
| `llama3_70B-pretrain.yaml` | `ici_fsdp_parallelism: 8`, `ici_data_parallelism: 1`, `dcn_fsdp_parallelism: 1`, `dcn_data_parallelism: -1` |
| `llama3_8B-pretrain.yaml` | `ici_fsdp_parallelism: 8`, `ici_data_parallelism: 1`, `dcn_fsdp_parallelism: 1`, `dcn_data_parallelism: -1` |
| `mixtral_8x7B-pretrain.yaml` | `ici_fsdp_parallelism: 1`, `ici_data_parallelism: 1`, `dcn_fsdp_parallelism: 1`, `dcn_data_parallelism: -1` |
| `qwen3_14B-pretrain.yaml` | `ici_fsdp_parallelism: 8`, `ici_data_parallelism: 1`, `dcn_fsdp_parallelism: 1`, `dcn_data_parallelism: -1` |
| `qwen3_30B_A3B-pretrain.yaml` | `ici_fsdp_parallelism: 1`, `ici_data_parallelism: 1`, `dcn_fsdp_parallelism: 1`, `dcn_data_parallelism: -1` |

The `llama2_7B-pretrain.yaml` example also sets `dataset_type: "synthetic"` and `hf_access_token: ${HF_TOKEN:""}` for gated Hugging Face assets when you switch to real data.

---

## Common patterns

### Testing with mock data

> Set `mock_data: true` (Megatron/TorchTitan) or synthetic dataset settings (MaxText) to validate configs and infrastructure without I/O-heavy datasets.

Reviewer: ```suggestion
Set `mock_data: true` (Megatron/TorchTitan) or synthetic dataset settings (MaxText) to validate the configurations and infrastructure without I/O-heavy datasets.
```

### Real training data

> - Megatron: configure `train_data_path` / `valid_data_path` and tokenizer assets in `overrides` once `mock_data` is false.  

Reviewer: ```suggestion
- Megatron: Configure `train_data_path` / `valid_data_path` and tokenizer assets in `overrides` once `mock_data` is false.  
```

> - Ensure host paths are mounted in **container** mode (`--volume` or `container.options.volume` in YAML).  

Reviewer: Does this apply to all backends? Or to which one?

> - TorchTitan / MaxText: follow backend-specific dataset fields in the experiment `overrides` and presets.

Reviewer: ```suggestion
- TorchTitan/MaxText: Follow backend-specific dataset fields in the experiment `overrides` and presets.
```

### Scaling from single-node to multi-node

> - Use **Slurm** mode for allocation; keep **container** entry if you want the same image on every node.  

Reviewer: ```suggestion
- Use **Slurm** mode for allocation; keep the **container** entry if you want the same image on every node.  
```

- Set environment variables consistently (`NNODES`, `NODE_RANK`, `MASTER_ADDR`, `MASTER_PORT`, `GPUS_PER_NODE`); the Slurm entry script injects them when using `primus-cli slurm`.  

> - Increase parallelism fields (Megatron TP/PP/EP; TorchTitan `parallelism`; MaxText `ici_*` / `dcn_*`) to match topology.

Reviewer: ```suggestion
- Increase values in the parallelism fields (Megatron TP/PP/EP; TorchTitan `parallelism`; MaxText `ici_*` / `dcn_*`) to match topology.
```

### Hugging Face token for gated models

> Export `HF_TOKEN` on the host before launching **container** mode; `runner/.primus.yaml` lists `HF_TOKEN` under `container.options.env` so it can be forwarded into the container. MaxText configs may reference `${HF_TOKEN:""}` directly.

Reviewer: ```suggestion
Export `HF_TOKEN` on the host before launching **container** mode; `runner/.primus.yaml` lists `HF_TOKEN` under `container.options.env` so it can be forwarded into the container. MaxText configurations may reference `${HF_TOKEN:""}` directly.
```

### HipBLASLt autotuning (three stages)

Controlled with `PRIMUS_HIPBLASLT_TUNING_STAGE` (see `examples/README.md`):

| Stage | Purpose |
| --- | --- |
| 1 | Dump GEMM shapes seen during training (reduce `train_iters` for faster collection). |
| 2 | Tune kernels from dumped shapes (offline tooling under `examples/offline_tune`). |
| 3 | Train using tuned kernel artifacts from `./output/tune_hipblaslt/...`. |

Example (from in-repo docs):

```bash
export PRIMUS_HIPBLASLT_TUNING=1        # master switch (required; tuning is skipped without it)
export PRIMUS_HIPBLASLT_TUNING_STAGE=1
./runner/primus-cli direct -- train pretrain \
  --config examples/megatron/configs/MI300X/llama2_7B-BF16-pretrain.yaml
```

---

> ## Supported models (summary)

Reviewer: ```suggestion
## Full list of supported models
```

> The tables in the Megatron, TorchTitan, and MaxText sections are curated MI300X examples from this repository. Use the filesystem under `examples/<backend>/configs/` as the authoritative inventory because new presets and hardware-specific examples may be added without a full rewrite of this guide.

Reviewer: ```suggestion
The tables above in the Megatron, TorchTitan, and MaxText sections are curated MI300X examples from the [Primus repository](https://github.com/AMD-AGI/Primus). Use `examples/<backend>/configs/` in the repository as the authoritative inventory, as new presets and hardware-specific examples may be added there before this document is updated to reflect their additions.
```

Reviewer: Do not use "this repository" as if the readers are reading the page inside the repo. Treat this as a standalone doc.
